File: src/main/java/com/rimlang/rim/translation/Translator.java
package com.rimlang.rim.translation;

import com.rimlang.rim.Main;
import com.rimlang.rim.lexer.NumberToken;
import com.rimlang.rim.lexer.StringToken;
import com.rimlang.rim.lexer.Token;
import com.rimlang.rim.lexer.TokenType;
import com.rimlang.rim.syntax.*;
import com.rimlang.rim.util.Strings;
import me.zeepic.JavaFormatter;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

public class Translator {

    private static final List<String> imports = new ArrayList<>();
    private static final List<String> events = new ArrayList<>();
    private static final List<String> methods = new ArrayList<>();

    private static final List<String> fields = new ArrayList<>();

    private static final StringBuilder onEnable = new StringBuilder();

    private static final HashMap<String, HashMap<String, String>> traits = new HashMap<>(); // <class, <trait_name, type>>

    public static void translateGlobalScope(List<SyntaxNode> nodes) {
        String eventMapString = Main.loadResource("event_map.txt");
        HashMap<String, String> eventMap = new HashMap<>();
        assert eventMapString != null;
        for (String line : eventMapString.split("\n")) {
            String[] split = line.split(": ");
            eventMap.put(split[1], split[0]);
        }
        HashMap<String, Integer> eventCount = new HashMap<>();
        boolean hasTimer = false;
        for (SyntaxNode node : nodes) {
            if (node instanceof ComplexFunctionCallSyntaxNode) {
                var syntaxNode = (ComplexFunctionCallSyntaxNode) node;
                switch (syntaxNode.getFunctionIdentifier().getValue()) {
                    case "on" -> {
                        String event = syntaxNode.getArgs().get(0).getTokens().stream().map(t -> {
                            if (t instanceof StringToken) {
                                return ((StringToken) t).getValue();
                            }
                            else return "";
                        }).collect(Collectors.joining(""));
                        if (event.equals("server.start")) {
                            onEnable.append(translate(syntaxNode.getBody()));
                        } else {
                            String spigotEvent = eventMap.getOrDefault(event, "");
                            if (spigotEvent.isEmpty()) {
                                throw new IllegalArgumentException("Unknown event: " + event + " (at line " + syntaxNode.getArgs().get(0).getTokens().get(0).getLine() + ").");
                            }
                            if (!eventCount.containsKey(event)) {
                                eventCount.put(event, 0);
                            } else {
                                eventCount.put(event, eventCount.get(event) + 1);
                            }
                            String count = eventCount.get(event) + "";
                            if (count.equals("0")) count = "";
                            events.add("@EventHandler\npublic void on" + spigotEvent + count + "(" + spigotEvent + " event) {\n" + translate(syntaxNode.getBody()) + "\n}");
                        }
                    }
                    case "trait" -> {
                        Token traitType = syntaxNode.getArgs().get(0).getTokens().get(0);
                        if (traitType instanceof StringToken) {
                            String traitName = ((StringToken) traitType).getValue();
                            if (!traits.containsKey(traitName)) {
                                traits.put(traitName, new HashMap<>());
                            }
                            for (SyntaxNode bodyNode : syntaxNode.getBody()) {
                                if (bodyNode instanceof VariableDefinitionSyntaxNode) {
                                    var variableDefinition = (VariableDefinitionSyntaxNode) bodyNode;
                                    traits.get(traitName)
                                            .put(variableDefinition.getIdentifier().getValue(), variableDefinition.getInitialValue().getResultType());
                                }
                            }
                        }
                    }
                    case "timer" -> {
                        if (!hasTimer) {
                            hasTimer = true;
                            methods.add("""
                                    private void beginTimer(double seconds, Runnable runnable) {
                                        Bukkit.getScheduler().runTaskTimer(this, runnable, 0, (long) (20 * seconds));
                                    }""".indent(4));
                        }
                        Token secondsToken = syntaxNode.getArgs().get(0).getTokens().get(0);
                        if (secondsToken instanceof NumberToken) {
                            NumberToken numberToken = (NumberToken) secondsToken;
                            onEnable.append("beginTimer(")
                                    .append(numberToken.getValue())
                                    .append(", () -> {\n")
                                    .append(translate(syntaxNode.getBody()))
                                    .append("\n});\n");
                        } else {
                            throw new IllegalArgumentException("Timer's first argument for seconds must be a number (at line " + secondsToken.getLine() + ").");
                        }
                    }
                }
            } else if (node instanceof CommandDefinitionSyntaxNode) {
                var syntaxNode = (CommandDefinitionSyntaxNode) node;
                // TODO: Implement command arguments
                onEnable.append("getCommand(\"")
                        .append(getCamelCase(syntaxNode.getIdentifier().getValue()))
                        .append("\").setExecutor((sender, command, label, args) -> {\n")
                        .append(translate(syntaxNode.getBody()))
                        .append("\n});");
            } else if (node instanceof FunctionDefinitionSyntaxNode) {
                var syntaxNode = (FunctionDefinitionSyntaxNode) node;
                methods.add("private " + toJavaType(syntaxNode.getReturnType()) + " " + getCamelCase(syntaxNode.getIdentifier().getValue()) + "(" + syntaxNode.getArgs().translate() + ") {\n" + translate(syntaxNode.getBody()) + "\n}");
            } else if (node instanceof VariableDefinitionSyntaxNode) {
                var syntaxNode = (VariableDefinitionSyntaxNode) node;
                fields.add("private var " + getCamelCase(syntaxNode.getIdentifier().getValue()) + " = " + syntaxNode.getInitialValue().translate() + ";");
            } else {
                throw new IllegalStateException("Node type " + node.getClass().getName() + " is not allowed in global scope.");
            }
        }
    }

    public static String toJavaType(StringToken typeToken) {
        String type = typeToken.getValue();
        if (type.startsWith("list[") && type.endsWith("]")) {
            return "List<" + toJavaType(new StringToken(
                    TokenType.TYPE,
                    type.substring(5, type.length() - 1),
                    typeToken.getLine(),
                    typeToken.getCharacter()
            )) + ">";
        }
        switch (type) {
            case "str":
                return "String";
            case "num":
                return "double";
            case "bool":
                return "boolean";
            case "void":
                return "void";
            case "player":
                return "Player";
        }
        try {
            throw new ClassNotFoundException("Type " + type + " not found (on line " + typeToken.getLine() + ").");
        } catch(ClassNotFoundException e) {
            e.printStackTrace();
            return "";
        }
    }

    public static String getCamelCase(String value) {
        String pascalCase = getPascalCase(value);
        return pascalCase.substring(0, 1).toLowerCase() + pascalCase.substring(1);
    }

    @NotNull
    private static String getPascalCase(String value) {
        return Arrays.stream(value.split("_"))
                .map(Strings::title)
                .collect(Collectors.joining());
    }

    public static String translate(List<SyntaxNode> nodes) {
        StringBuilder builder = new StringBuilder();
        for (SyntaxNode node : nodes) {
            if (node instanceof IfSyntaxNode) {
                var syntaxNode = (IfSyntaxNode) node;
                builder.append("if (")
                        .append(syntaxNode.getBooleanExpression().translate())
                        .append(") {\n")
                        .append(translate(syntaxNode.getBody()))
                        .append("\n}\n");
            } else if (node instanceof ExpressionSyntaxNode) {
                var syntaxNode = (ExpressionSyntaxNode) node;
            } else if (node instanceof ForeachSyntaxNode) {
                var syntaxNode = (ForeachSyntaxNode) node;
                builder.append("for (");
            } else if (node instanceof WhileSyntaxNode) {
                var syntaxNode = (WhileSyntaxNode) node;
            } else if (node instanceof ReturnSyntaxNode) {
                var syntaxNode = (ReturnSyntaxNode) node;
            } else if (node instanceof VariableDefinitionSyntaxNode) {
                var syntaxNode = (VariableDefinitionSyntaxNode) node;
            }
        }
        return builder.toString();
    }

    public static String complete(String pluginName) {
        String template = Main.loadResource("Template.java");
        assert template != null;
        String complete = template.formatted(
                String.join(";\nimport ", imports),
                pluginName,
                String.join("\n", fields),
                onEnable.toString(),
                String.join("\n", events),
                String.join("\n\n", methods)
        );
        return JavaFormatter.format(complete);
    }

}

File: src/main/java/com/rimlang/rim/translation/Translator.kt
package com.rimlang.rim.translation

import com.rimlang.rim.asResource
import com.rimlang.rim.lexer.NumberToken
import com.rimlang.rim.lexer.StringToken
import com.rimlang.rim.lexer.TokenType
import com.rimlang.rim.syntax.*
import com.rimlang.rim.util.camelCase
import com.rimlang.rim.util.classNameFromQualifiedName
import com.rimlang.rim.util.pascalCase
import org.eclipse.jdt.core.JavaCore
import org.eclipse.jdt.core.ToolFactory
import org.eclipse.jdt.core.formatter.CodeFormatter
import org.eclipse.jdt.core.formatter.DefaultCodeFormatterConstants
import org.eclipse.jface.text.Document


private fun Int.toStringWithoutZero()
    = if (this == 0) ""
      else this.toString()

object Translator {

    private val imports = mutableListOf<String>()
    private val events = mutableListOf<String>()
    private val methods = mutableListOf<String>()
    private val fields = mutableListOf<String>()
    private var onEnable = ""
    val traits = mutableMapOf<String, MutableMap<String, String>>() // <class, <trait_name, type>>

    private val eventMap = HashMap<String, String>()


    private val globalContext = GlobalContext()

    init {
        val eventMapString = "event_map.txt".asResource()

        // Create event map
        for (line in eventMapString.split("\n")) {
            val split = line.split(": ")
            if (split.size < 2) continue
            eventMap[split[1].trim()] = split[0]
        }
        println(eventMap)
    }

    fun translateGlobalScope(nodes: List<SyntaxNode>) {
        val eventCount = HashMap<String, Int>()
        var hasTimer = false
        nodes.filterIsInstance<FunctionDefinitionSyntaxNode>()
            .forEach {
                val id = it.identifier.value.camelCase()
                methods += "private ${toJavaType(it.returnType)} $id(${it.args.translate(globalContext)}) {"
                val variableArgs = it.args.args.map { arg ->
                    VariableIdentifier(arg.identifier.value, arg.type.value, "", arg.identifier.line)
                }
                methods += translate(it.body, FunctionContext(id,
                    (globalContext.identifiers + variableArgs).toMutableList()
                ))
                methods += "}\n"
                globalContext.identifiers += FunctionIdentifier(id,
                    it.args
                        .args
                        .map { arg -> arg.translate().split(" ") }
                        .associate { (a, b) -> a to b },
                    it.returnType.value,
                    it.identifier.line
                )
            }
        for (node in nodes) {
            when (node) {
                is ComplexFunctionCallSyntaxNode -> {
                    when (node.functionIdentifier.value) {
                        "on" -> {
                            val event = node.args[0].tokens
                                .filterIsInstance<StringToken>()
                                .joinToString("") { it.value }
                            if (event == "server.start") {
                                onEnable += translate(node.body, EventContext("onEnable", globalContext.identifiers))
                            } else {
                                val spigotEvent = eventMap[event] ?: ""
                                require(spigotEvent.isNotEmpty()) {
                                    "Unknown event: $event (at line ${node.args[0].tokens[0].line})."
                                }
                                eventCount[event] = if (event !in eventCount) 0
                                                    else eventCount[event]!! + 1
                                val count = eventCount[event]!!.toStringWithoutZero()
                                events += "@EventHandler"
                                events += "public void on$spigotEvent$count(${spigotEvent.classNameFromQualifiedName()} event) {"
                                events += translate(node.body, EventContext(spigotEvent, globalContext.identifiers))
                                events += "}"
                            }
                        }
                        "trait" -> {
                            val traitType = node.args[0].tokens[0] as StringToken
                            val traitName = traitType.value
                            if (traitName !in traits) {
                                traits[traitName] = mutableMapOf()
                            }
                            node.body.filterIsInstance<VariableDefinitionSyntaxNode>()
                                .forEach {
                                    it.initialValue.translate(globalContext)
                                    traits[traitName]!![it.identifier.value] = it.initialValue.resultType ?: "String"
                                }
                        }
                        "timer" -> {
                            if (!hasTimer) {
                                hasTimer = true
                                methods += "    private void beginTimer(double seconds, Runnable runnable) {"
                                methods += "        Bukkit.getScheduler().runTaskTimer(this, runnable, 0, (long) (20 * seconds));"
                                methods += "    }"
                            }
                            val secondsToken = node.args[0].tokens[0]
                            require(secondsToken is NumberToken) {
                                "Timer argument must be a number (at line ${secondsToken.line})."
                            }
                            onEnable += "beginTimer(${secondsToken.value}, () -> {\n"
                            onEnable += translate(node.body, TimerContext(globalContext.identifiers))
                            onEnable += "\n});\n"
                        }
                    }
                }
                is CommandDefinitionSyntaxNode -> {
                    // TODO: Implement command arguments
                    onEnable += "getCommand(\"${node.identifier.value.camelCase()}\""
                    onEnable += ").setExecutor((sender, command, label, args) -> {\n"
                    onEnable += translate(node.body, CommandContext(globalContext.identifiers))
                    onEnable += "});\n"
                }
                is FunctionDefinitionSyntaxNode -> continue
                is VariableDefinitionSyntaxNode -> {
                    fields += "private var ${node.identifier.value.pascalCase()} = ${
                        node.initialValue.translate(VariableDefinitionContext(globalContext.identifiers))
                    };"
                    globalContext.identifiers += VariableIdentifier(
                        node.identifier.value,
                        node.initialValue.resultType ?: throw IllegalStateException("Variable must be initialized with a real value!"),
                        node.initialValue.translate(globalContext),
                        node.identifier.line
                    )
                }
                else -> {
                    throw IllegalStateException("Node type " + node::class.simpleName + " is not allowed in global scope.")
                }
            }
        }
    }

    fun toJavaType(typeToken: StringToken): String {
        val type = typeToken.value
        if (type.startsWith("list[") && type.endsWith("]")) {
            return "List<" + toJavaType(
                StringToken(
                    TokenType.TYPE,
                    type.substring(5, type.length - 1),
                    typeToken.line,
                    typeToken.character
                )
            ) + ">"
        }
        when (type) {
            "str" -> return "String"
            "num" -> return "double"
            "bool" -> return "boolean"
            "void" -> return "void"
            "player" -> return "Player"
        }
        throw ClassNotFoundException("Type " + type + " not found (on line " + typeToken.line + ").")
    }

    private fun translate(nodes: List<SyntaxNode>, context: Context): String {
        var builder = ""
        for (node in nodes) {
            when (node) {
                is IfSyntaxNode -> {
                    node.branches.forEach {
                        val value = it.type.value.takeUnless { v -> v in TokenType.ELIF.values } ?: "else if"
                        builder += "$value ${if (value != "else") "(${it.booleanExpression.translate(context)})" else ""} {"
                        builder += translate(it.body, context)
                        builder += "} "
                    }
                }
                is ExpressionSyntaxNode -> {
                    println(node)
                }
                is ForeachSyntaxNode -> {
                    val loopType = node.genericExpression.resultType ?: "Object"
                    builder += "for ($loopType ${node.loopIdentifier.value.camelCase()} : ${node.genericExpression.translate(ForeachContext(loopType, globalContext.identifiers))}) {\n"
                    builder += translate(node.body, context)
                    builder += "}\n"
                }
                is WhileSyntaxNode -> {
                }
                is ReturnSyntaxNode -> {
                }
                is VariableDefinitionSyntaxNode -> {
                    builder += "var ${node.identifier.value.pascalCase()} = ${
                        node.initialValue.translate(VariableDefinitionContext(globalContext.identifiers))
                    };\n"
                    context.identifiers += VariableIdentifier(
                        node.identifier.value,
                        node.initialValue.resultType ?: throw IllegalStateException("Variable must be initialized with a real value!"),
                        node.initialValue.translate(globalContext),
                        node.identifier.line
                    )
                }
            }
        }
        return builder
    }

    fun complete(pluginName: String): String {
        val template = "Template.java".asResource()
        val complete = template.format(
            imports.joinToString(";\nimport "),
            pluginName,
            fields.joinToString("\n"),
            onEnable,
            events.joinToString("\n"),
            methods.joinToString("\n")
        )
        val options = DefaultCodeFormatterConstants.getEclipseDefaultSettings()

        options[JavaCore.COMPILER_COMPLIANCE] = JavaCore.VERSION_1_5
        options[JavaCore.COMPILER_CODEGEN_TARGET_PLATFORM] = JavaCore.VERSION_1_5
        options[JavaCore.COMPILER_SOURCE] = JavaCore.VERSION_1_5

        // Can use: DefaultCodeFormatterConstants.SOMETHING
        val formatter = ToolFactory.createCodeFormatter(options)
        val edits = formatter.format(CodeFormatter.K_COMPILATION_UNIT, complete, 0, complete.length, 0, "\n")
        val document = Document(complete)
        edits.apply(document)
        return document.get()
    }
}